lib/sinatra/base.rb

```diff
@@ -1023,13 +1023,13 @@ def process_route(pattern, conditions, block = nil, values = [])
       return unless params = pattern.params(route)
 
       params.delete("ignore") # TODO: better params handling, maybe turn it into "smart" object or detect changes
-      original, @params = @params, @params.merge(params) if params.any?
+      original, @params = @params, @params.merge(force_encoding(params)) if params.any?
 
       regexp_exists = pattern.is_a?(Mustermann::Regular) || (pattern.respond_to?(:patterns) && pattern.patterns.any? {|subpattern| subpattern.is_a?(Mustermann::Regular)} )
       if regexp_exists
         captures           = pattern.match(route).captures
         values            += captures
-        @params[:captures] = captures unless captures.nil? || captures.empty?
+        @params[:captures] = force_encoding(captures) unless captures.nil? || captures.empty?
       else
         values += params.values.flatten
```

Comment:
Is it also necessary to handle this line to properly encode the block arguments as follows?
```ruby
get '/:id' |id|
  "#{id.encoding}"
end
```

Reply:
I don't believe it's necessary to change that line.  force_encoding operates recursively in place, so if there were any items to change the encoding of, we'd have already done it on line 1026.  I agree this behavior of force_encoding is surprising, but it's probably that way for efficiency reasons.

I did add a test for this case below to demonstrate that it works.

I suppose to make it more obvious, I could hoist the force_encoding above line 1026 and put it in void context.  Let me make that change, as I think it would make things easier to understand.

Reply:
It makes sense. Thanks for your explanation.